homeassistant/helpers/service.py

```diff
@@ -113,3 +117,58 @@ def extract_entity_ids(hass, service_call):
         return group.expand_entity_ids(hass, [service_ent_id])
 
     return [ent_id for ent_id in group.expand_entity_ids(hass, service_ent_id)]
+
+
+# pylint: disable=too-few-public-methods
+class CallRateDelayThrottle(object):
+    """Helper class to provide service call rate throttling.
+
+    This class provides a decorator to decorate service methods that need
+    to be throttled to not exceed a certain call rate per second.
+    One instance can be used on multiple service methods to archive
+    an overall throttling.
+
+    As this uses track_point_in_utc_time to schedule delayed executions
+    it should be async friendly.
```

Comment:
OK.  What do I have to do to make it async friendly?  Just change it to the async_track_point_in_time version and remove the lock? As long as it is guaranteed that services are only called through the main loop the lock should not be necessary.